### PR TITLE
add reports making analysis easier

### PR DESCRIPTION
**Pull request by derhuerst**
related: #33

### CODE REVIEW

File: docs/analysis/feed-by-route-date.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+- `date`
+- `dow` – day of the week, following the [PostgreSQL notation `0` (Sunday) to `6` (Saturday)](https://www.postgresql.org/docs/14/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT)
+- `nr_of_trips` – nr of trips starting on that date
+- `nr_of_arrs_deps` – nr of trips taking place on that date
+- `is_effective` – wether `nr_of_trips` & `nr_of_arrs_deps` are calculated based on the *effective* date (i.e. the date that the stop time event actually happens on) or *schedule* date (i.e. the date which their `stop_time` rows refer to)
+
+So
+
+- if you want to take a customer-facing perspective on the data (as in "I don't care which if trips are scheduled before midnight, I want to know if they run today"), filter for `is_effective = True` rows;
```
Comment (hbruch) [inline docs/analysis/feed-by-route-date.md:24]: typo `which if`?

File: lib/stats_by_agency_route_stop_hour.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+FROM "${opt.schema}".arrivals_departures;
+
+${materialized ? `\
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (stop_id);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (effective_hour);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id, stop_id);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id, effective_hour);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (stop_id, effective_hour);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id, stop_id, effective_hour);
```
Comment (hbruch) [inline lib/stats_by_agency_route_stop_hour.js:30]: I'd assume stop_id to be more specific than route_id, and give reordering a try (using stop_id, route_id, effective_hour)

File: lib/stats_by_route_date.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+const afterAll = (opt) => {
+	let materialized = false
+	if (opt.statsByRouteIdAndDate === 'materialized-view') {
+		materialized = true
+	} else if (opt.statsByRouteIdAndDate !== 'view') {
+		throw new Error('invalid opt.statsByRouteIdAndDate, must be one of these: none, view, materialized-view.')
+	}
+	const createViewCmd = materialized
+		? `CREATE MATERIALIZED VIEW`
+		: `CREATE OR REPLACE VIEW`
```
Comment (hbruch) [inline lib/stats_by_route_date.js:12]: `CREATE OR REPLACE VIEW` looks like idempotency intended, while for `MATERIALIZED VIEW`,a preexisting view is not dropped(?)

File: lib/stats_by_agency_route_stop_hour.js (hunk 2 above)
Comment (derhuerst) [reply lib/stats_by_agency_route_stop_hour.js:30]: With the 2023-05-02 VBB GTFS, the indexes defined above take up ~7GB, while the materialised view itself takes up ~10GB.

The [PostgreSQL docs say this](https://www.postgresql.org/docs/15/indexes-multicolumn.html):

> A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

However, it seems like having (a) multi-column index(es) is not necessary, as filtering by one of the columns (`route_id` or `stop_id` or `station_id` or `effective_hour`) reduces the number of rows enough (down to 10^4 - 10^5) for queries to be fast enough (<= 10ms). Defining one single-column index for each column yields ~3GB of indexes.

What do you think?

File: lib/stats_by_route_date.js (hunk 3 above)
Comment (derhuerst) [reply lib/stats_by_route_date.js:12]: True, in order to stay simple, `gtfs-via-postgres` is *not* intended for idempotency. I just followed the rest of the code base here. I will adapt all the occurences eventually.

**Comment by derhuerst** [comment]
FYI: I have added another view `stats_active_trips_by_hour` that focuses on a slightly different perspective than `stats_by_{route_date,agency_route_stop_hour}`: Instead of computing the number of arrivals/departures within a time frame, it calculates the number of *currently running* trips at any point within each hour.

**Pull request merged**